### PR TITLE
Prevent running a ClusterDeprovision job if owning cluster is not del…

### PR DESCRIPTION
**Pull request by dgoodwin**
…eted.

### CODE REVIEW

**Comment by dgoodwin** [comment]
Adds a check to make sure the ClusterDeployment (should one exist) that owns a ClusterDeprovosion has been deleted. Prevents an unlikely case where someone created a ClusterDeprovision manually.

Not sure if this makes sense, I'm more worried about something mistakenly issuing a ClusterDeployment delete tbh.

**Comment by dgoodwin** [comment]
Also InstallManager also has a guard against running if the ClusterDeployment is set installed, so the provision case should be protected.

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 1]
```diff
@@ -135,6 +137,25 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
 		return reconcile.Result{}, nil
 	}
 
+	// Check if there is a ClusterDeployment owning this Deprovision, if so look it up and
+	// make sure it has a deletion timestamp. Otherwise bail out as a safety check.
+	for _, oRef := range instance.OwnerReferences {
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: `GetControllerOf` from "k8s.io/apimachinery/pkg/apis/meta/v1"

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 2]
```diff
@@ -135,6 +137,25 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
 		return reconcile.Result{}, nil
 	}
 
+	// Check if there is a ClusterDeployment owning this Deprovision, if so look it up and
+	// make sure it has a deletion timestamp. Otherwise bail out as a safety check.
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: The controller should also bail out if the deprovision is **not owned** by a ClusterDeployment.

Comment (dgoodwin) [reply pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: I wasn't sure where to go with that as technically at one point this was designed and used to be able to cleanup clusters that were provisioned with hive and the preserveOnDelete option, then later deprovision them without there being a cluster deployment. Feature was axed in OCM but might make a reappearance so it was theoretically possible that there could be a detached unassociated Deprovision.

Comment (staebler) [reply pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: I'm not keen on being in the business of providing a means to delete a random cluster not controller by Hive. I would rather prevent the deletion now and relax the restriction in the future if we decide to explicitly allow the feature.

Comment (dgoodwin) [reply pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: Deal.

Comment (staebler) [reply pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: Not thinking about it too much, but I think I would feel better if, to cover the use case described, the user had to create a new clusterdeployment that adopted the existing cluster.

**Comment by staebler** [comment]
@dgoodwin Should this go in master?

**Comment by dgoodwin** [comment]
Pretty grey, but I think v1 is ok, it's not critical to get it out to master and I mostly want it there when we get to migration work.

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 3]
```diff
@@ -135,6 +137,32 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
...
+	oRef := metav1.GetControllerOf(instance)
+	if oRef == nil {
+		// TODO: this was once supported to cleanup self-managed "preserveOnDelete" clusters,
+		// but the feature was killed off. For now we'd rather not open the door to dangling
+		// ClusterDeprovisions with no associated cluster.
+		rLog.Warn("ClusterDeprovision does not have an owning ClusterDeployment")
+		return reconcile.Result{}, fmt.Errorf("ClusterDeprovision does not have an owning ClusterDeployment")
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: This should not return an error. There is no reason to resync the resource.

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 4]
```diff
@@ -135,6 +137,32 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
...
+	}
+	if oRef.Kind != "ClusterDeployment" || !strings.HasPrefix(oRef.APIVersion, "hive.openshift.io") {
+		rLog.Warnf("ClusterDeprovision has a non-ClusterDeployment owner: %v", oRef)
+		return reconcile.Result{}, fmt.Errorf("ClusterDeprovision has a non-ClusterDeployment owner: %v", oRef)
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: Same thing about not returning an error.

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 5]
```diff
@@ -135,6 +137,32 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
...
+	}
+	cd := &hivev1.ClusterDeployment{}
+	err = r.Get(context.TODO(), types.NamespacedName{Namespace: instance.Namespace, Name: oRef.Name}, cd)
+	if err != nil {
+		rLog.Error("error looking up ClusterDeployment that owns ClusterDeprovision")
+		return reconcile.Result{}, nil
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: This **should** return an error.

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 6]
```diff
@@ -135,6 +137,32 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
...
+	}
+	if cd.DeletionTimestamp == nil {
+		err = fmt.Errorf("ClusterDeprovision created for ClusterDeployment that has not been deleted")
+		rLog.WithError(err).Error("skipping")
+		return reconcile.Result{}, err
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: This should not return an error.

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 7]
```diff
@@ -135,6 +137,32 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
...
+	}
+	cd := &hivev1.ClusterDeployment{}
+	err = r.Get(context.TODO(), types.NamespacedName{Namespace: instance.Namespace, Name: oRef.Name}, cd)
+	if err != nil {
+		rLog.Error("error looking up ClusterDeployment that owns ClusterDeprovision")
+		return reconcile.Result{}, nil
+	}
+	if cd.DeletionTimestamp == nil {
+		err = fmt.Errorf("ClusterDeprovision created for ClusterDeployment that has not been deleted")
+		rLog.WithError(err).Error("skipping")
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: I don't see any need to use `WithError` here. Make the message of the log entry the text of what you have as the error.

File: pkg/controller/clusterdeprovision/clusterdeprovision_controller.go [hunk 8]
```diff
@@ -135,6 +137,32 @@ func (r *ReconcileClusterDeprovision) Reconcile(request reconcile.Request) (reco
...
+		// ClusterDeprovisions with no associated cluster.
+		rLog.Warn("ClusterDeprovision does not have an owning ClusterDeployment")
+		return reconcile.Result{}, fmt.Errorf("ClusterDeprovision does not have an owning ClusterDeployment")
+	}
+	if oRef.Kind != "ClusterDeployment" || !strings.HasPrefix(oRef.APIVersion, "hive.openshift.io") {
+		rLog.Warnf("ClusterDeprovision has a non-ClusterDeployment owner: %v", oRef)
+		return reconcile.Result{}, fmt.Errorf("ClusterDeprovision has a non-ClusterDeployment owner: %v", oRef)
+	}
+	cd := &hivev1.ClusterDeployment{}
+	err = r.Get(context.TODO(), types.NamespacedName{Namespace: instance.Namespace, Name: oRef.Name}, cd)
```
Comment (staebler) [inline pkg/controller/clusterdeprovision/clusterdeprovision_controller.go]: `if err := ...; err != nil {` ?

**Comment by dgoodwin** [comment]
Updated.

**Pull request merged**